app/controllers/users_controller.rb

```diff
@@ -0,0 +1,91 @@
+class UsersController < ApplicationController
+  layout 'index'
+
+  before_action :set_user, only: [ :show, :edit, :update, :destroy, :delete ]
+  before_action :require_login, only: [ :edit, :update, :destroy, :delete ]
+
+  def index
+    @user = User.new
+    @users = User.all
+  end
+
+  def show
+    @user
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      redirect_to sign_in_users_path, notice: 'User created successfully! Please sign in.'
+    else
+      flash.now[:alert] = @user.errors.full_messages.join(', ')
+      render :new
+    end
+  end
+
+  def sign_in
+    @user = User.new
+  end
+
+  def create_session
+    user = User.find_by(email: params[:email])
+    if user&.authenticate_the_login(params[:password])
+      session[:user_id] = user.id
+      redirect_to products_path, notice: 'Successfully signed in!'
+    else
+      flash.now[:alert] = 'Invalid email or password'
+      render :sign_in
+    end
+  end
+
+  def destroy_session
+    session[:user_id] = nil
+    redirect_to sign_in_users_path, notice: 'Successfully signed out!'
+  end
+
+  def edit
+    @user = User.find(params[:id])
+  end
+
+  def update
+    @user = User.find(params[:id])
+    if @user.update(user_params)
+      redirect_to products_path, notice: 'User updated successfully!'
+    else
+      render :edit
+    end
+  end
+
+  def delete
+  end
+
+  def destroy
+    @user = User.find(params[:id])
+    @user.destroy
+    flash[:notice] = 'User deleted successfully.'
+    redirect_to users_path
+  end
+
+
+  def products
+  end
+
```

Comment:
What is the purpose of this method?

Reply:
There isn't really any purpose but I added it as I needed to have this  to go from the Users later.

Reply:
Let's remove codes that is not related to the works of this PR.
Also, will you be merging authentication without gem?

Reply:
Okay Apu, removing it. Thank you.